Add API to create funded raw transactions

dexX7 opened the pull request:
This pull request adds two new RPCs "omni_createfunded_send" and "omni_createfunded_sweep", which allow the creation of raw Omni Layer transactions, which are funded by a different source, which isn't necessarily the sender.

This can be used, when the sender only has a tiny fraction of coins available, but not enough to cover the fees of the transaction.

**All coins from the specified sender are consumed and used and if coins are missing to create the transaction, they are taken from the specified fee source. Change is sent to the fee source!**

**As per default, the unspent outputs of this transaction are locked, to avoid that they are used for the construction of another transaction.**

The resulting raw transaction then needs to be signed and broadcasted.

This submission closes #383 and replaces #388 - at least to some degree.

TODO:

- [ ] update documentation

bigzhu commented:
Where I can checkout branch `dexX7:0.3.1-wallet-funded-send` ? 
I want test and use this two new RPCs for try

dexX7 commented:
Hi @bigzhu, this would be awesome and very helpful!

You can checkout them with:

```
git clone https://github.com/dexX7/bitcoin.git
git checkout 0.3.1-wallet-funded-send
```

Please let me know, how it went!

bigzhu commented:
It's not work. still show

```
./omnicore-cli omni_createfunded_send  mw7yogQhSTc2KmAB3DHZPdaHzWzR2Afdpj mrEkdJV5Pfer7VPHzdgdwsmVTXCPwqzsK6 1  0.1 mr9apo8ZazJs9iYnngwzFXmjsp9Fr74e3D
error code: -212
error message:
Error choosing inputs for the send transaction
```
`mw7yogQhSTc2KmAB3DHZPdaHzWzR2Afdpj` not have btc, but `mr9apo8ZazJs9iYnngwzFXmjsp9Fr74e3D` have btc

dexX7 commented:
> mw7yogQhSTc2KmAB3DHZPdaHzWzR2Afdpj not have btc, but mr9apo8ZazJs9iYnngwzFXmjsp9Fr74e3D have btc

You need *some* BTC, even if it's a tiny bit. This is because the first transaction input is used to determine the sender of a transaction.

bigzhu commented:
It's work fine, great work, thank your! 

When will merge to master?

```
./omnicore-cli omni_createfunded_send  mgxPp67ytTEEbTxkfwKH8CYgtFczxjYJ3d mr9apo8ZazJs9iYnngwzFXmjsp9Fr74e3D 1  0.3 mr9apo8ZazJs9iYnngwzFXmjsp9Fr74e3D
0100000002869aea40be5ee8fdf3c37a53f3894d49e9951d4e4ead6ec86776e7bcafeadbba0200000000ffffffff869aea40be5ee8fdf3c37a53f3894d49e9951d4e4ead6ec86776e7bcafeadbba0000000000feffffff036da7c404000000001976a914749eb9e8565f0678016b0397a2cb8b9db536d5c688ac0000000000000000166a146f6d6e6900000000000000010000000001c9c38022020000000000001976a914749eb9e8565f0678016b0397a2cb8b9db536d5c688ac00000000
dabank@localhost:~/omnicore_dev/src$ ./omnicore-cli "signrawtransaction" 0100000002869aea40be5ee8fdf3c37a53f3894d49e9951d4e4ead6ec86776e7bcafeadbba0200000000ffffffff869aea40be5ee8fdf3c37a53f3894d49e9951d4e4ead6ec86776e7bcafeadbba0000000000feffffff036da7c404000000001976a914749eb9e8565f0678016b0397a2cb8b9db536d5c688ac0000000000000000166a146f6d6e6900000000000000010000000001c9c38022020000000000001976a914749eb9e8565f0678016b0397a2cb8b9db536d5c688ac00000000
{
  "hex": "0100000002869aea40be5ee8fdf3c37a53f3894d49e9951d4e4ead6ec86776e7bcafeadbba020000006b483045022100d001c01cd88a3a1852b0993039396e2a384128099ed9dd111b19e443b24181e002201e596f8bf4606633be5709396c267780c2b5e1e9ef6c244372d7ba7f80cb34940121027b847e53417c92aea66c8e14cfbcf689b9b4d9d3c551bf5c98d155afb00e74dbffffffff869aea40be5ee8fdf3c37a53f3894d49e9951d4e4ead6ec86776e7bcafeadbba000000006b483045022100e49b7cddafc2303bb350c45aa2f50a418551e3f4233c6dc5acd7d363d43d8cce022042afa073e07f2012bd135a9f8ce3ababf888bbd4f460e38809b0119e917712770121027a31fd0526c98b6d0e74cc1c3b54dfcbbf272fc0b999f7509c664e32c7958403feffffff036da7c404000000001976a914749eb9e8565f0678016b0397a2cb8b9db536d5c688ac0000000000000000166a146f6d6e6900000000000000010000000001c9c38022020000000000001976a914749eb9e8565f0678016b0397a2cb8b9db536d5c688ac00000000",
  "complete": true
}
dabank@localhost:~/omnicore_dev/src$ ./omnicore-cli "sendrawtransaction" "0100000002869aea40be5ee8fdf3c37a53f3894d49e9951d4e4ead6ec86776e7bcafeadbba020000006b483045022100d001c01cd88a3a1852b0993039396e2a384128099ed9dd111b19e443b24181e002201e596f8bf4606633be5709396c267780c2b5e1e9ef6c244372d7ba7f80cb34940121027b847e53417c92aea66c8e14cfbcf689b9b4d9d3c551bf5c98d155afb00e74dbffffffff869aea40be5ee8fdf3c37a53f3894d49e9951d4e4ead6ec86776e7bcafeadbba000000006b483045022100e49b7cddafc2303bb350c45aa2f50a418551e3f4233c6dc5acd7d363d43d8cce022042afa073e07f2012bd135a9f8ce3ababf888bbd4f460e38809b0119e917712770121027a31fd0526c98b6d0e74cc1c3b54dfcbbf272fc0b999f7509c664e32c7958403feffffff036da7c404000000001976a914749eb9e8565f0678016b0397a2cb8b9db536d5c688ac0000000000000000166a146f6d6e6900000000000000010000000001c9c38022020000000000001976a914749eb9e8565f0678016b0397a2cb8b9db536d5c688ac00000000"
f186ca1e19fba9c03a9cf424ef099fb2295bf0800da0cb81f2c34ed4091e3737
```

Xavier-Lam commented:
my sender has only one unspent transaction output and 0.5usdt, i sent 0.3usdt to an address and spent the sender's transaction output, how can i send the rest 0.2 usdt? here is my transaction:
https://www.blockchain.com/btc/tx/0174665e310b87e0e1f4d24eacb7a837adb282561970d583f1d2777a4a5b31b5
there is no any transaction outputs for my sender to send the rest usdt, must i send some btcs to my sender?

dexX7 commented:
> When will merge to master?

It's going to be merged into develop in a few days and then we're going to release Omni Core 0.3.1 and at this point, it's going to be merged into master. I estimate this happens in roughly a month top.

> there is no any transaction outputs for my sender to send the rest usdt, must i send some btcs to my sender?

Yes, you have to send some BTC there.

dexX7 commented:
Related: #689

vetal20600 commented:
this is very useful. waiting on omnicore 0.3.1

igorbiliba commented:
how many long waiting update 0.3.1?

dexX7 commented:
> still have a tiny bit BTC for the first transaction input is used to determine the sender of a transaction ?

Correct. But this is usually the case, when you receive tokens.

dexX7 commented:
@all:

I modified the new RPCs as follows:

Instead of just creating a raw unsigned transaction, the new funding RPCs from #658 are modified to sign and broadcast the transactions in one step.

They are also renamed:

wll2007 commented:
@dexX7 now,I have a USDT address,only with USDT,but not have BTC. I want to send some USDT to the other who have a address. I create a funded raw transaction with my other address（the address with some BTC） ,is ok? if not ,how I send the transaction ?

dexX7 commented:
Hi @wll2007, you need at least a tiny bit of BTC. If you don't have any, I fear you have to send some there first.

wll2007 commented:
when merged to master 0.3.1 ?already  waite a long time !

DengMins commented:
@dexX7 There are 0.00000456 BTC in my account. Can I make a USTD transfer?

AllenYaoya commented:
> Hi @wll2007, you need at least a tiny bit of BTC. If you don't have any, I fear you have to send some there first.

Hi @dexX7, you mean at least a tiny bit of BTC is needed in the send address. what exactly is it?

dexX7 commented:
@AllenYaoya: Omni transactions are Bitcoin transactions and we need to send from a source to identify it as token sender. To send from, you need some Bitcoin.

hakerkiler commented:
Hi @dexX7,
omni_createfunded_send -> omni_funded_send
omni_createfunded_sweep -> omni_funded_sendall
This methods is not work on RPC omnicore
Exist new method?
